docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+
+### Deletion Changes
+
+While writes are de-conflicted by writers always using their own generation number in the key,
+deletions are slightly more challenging: if a pageserver A is isolated, and the true active node is
+pageserver B, then it is dangerous for A to do any object deletions, even of objects that it wrote
+itself, because pageserver's B metadata might reference those objects.
+
+To solve this without doing excessive synchronization, deletions are accumulated in deletion lists
+(also known as deadlists, or garbage lists), to be executed later. The actual execution of one of
+these lists has two preconditions:
+
+- That the executing node has written out the metadata index file since it decided to do
+  the deletion (i.e. no to-be-deleted files are referenced by the index)
+- That after writing that index file, the deleting node has confirmed that it is still the latest
+  generation (i.e. none of its attached tenants have been re-attached to another node).
+
+The order of operations for correctness is:
+
+1. Decide to delete one or more objects
+2. Write updated metadata (this means that if we are still the rightful leader, any subsequent leader will
+   read metadata that does not reference the about-to-be-deleted object).
+3. Confirm we hold a non-stale generation (this validates the assumption in step 2, and taken
+   together, these provide a guarantee that no future leader will attempt to read the objects
+   we are about to delete).
+4. Actually delete the objects
+
+Because step 3 puts load on the control plane, deletion lists should be executed very lazily: write
+them out to S3 and then execute them as a low priority batch operation in the background.
+
+Deletion lists have an a couple of additional benefits:
+
+- efficiency: accumulate enough objects
+  to populate plural DeleteObjects requests, rather than issuing many smaller `DeleteObject` requests.
+- read replicas: if some other node is reading data in S3 via some older metadata, the longer we delay
+  deletions in a deletion list, the less up-to-date we require that read replica to be when serving
+  reads for a historical LSN (i.e. for historical reads, leaving some stale objects around is
+  a good thing).
+
+There is some flexibility in exactly how deletion should work, but a simple implementation would look
+like this:
+
+- All Tenant objects hold a buffer of object keys to delete, which is flushed to the next step whenever
+  they write their index file (i.e. step 2 in the steps above)
+- On flush of the per-tenant buffer, keys to delete are gathered into a shared DeletionAccumulator,
+  which writes out deletion lists to S3 in batches of some size chosen for efficiency.
+- A background task wakes up and processes deletion lists. Before executing the actual deletion, it must call
+  into the control plane (step 3 in the steps above).
+
+Deletion lists in S3 may _not_ be executed as-is by future writers for the tenant: they
```

Comment:
It's not necessary for correctness.

Added text:
> Persisting
deletion lists to S3 is not strictly necessary for correctness (deletions can always be accomplished
later by object listing & scrub), but provides more flexibility in the case of a given node ID being
executed on a fresh node with an empty disk: it can recover the deletion queue from the previous
incarnation of the node ID.